vmm/src/device_manager.rs

```diff
@@ -611,47 +611,63 @@ impl DeviceTree {
 
 // Breadth first traversal iterator.
 struct BftIter<'a> {
-    next_layer: Vec<&'a DeviceNode>,
-    hash_map: &'a HashMap<String, DeviceNode>,
+    nodes: Vec<&'a DeviceNode>,
 }
 
 impl<'a> BftIter<'a> {
     fn new(hash_map: &'a HashMap<String, DeviceNode>) -> Self {
-        let mut next_layer = Vec::new();
+        let mut nodes = Vec::new();
 
         for (_, node) in hash_map.iter() {
             if node.parent.is_none() {
-                next_layer.push(node);
+                nodes.push(node);
             }
         }
 
-        BftIter {
-            next_layer,
-            hash_map,
+        let mut node_layer = nodes.as_slice();
+        loop {
+            let mut next_node_layer = Vec::new();
+
+            for node in node_layer.iter() {
+                for child_node_id in node.children.iter() {
+                    if let Some(child_node) = hash_map.get(child_node_id) {
+                        next_node_layer.push(child_node);
+                    }
+                }
+            }
+
+            if next_node_layer.is_empty() {
+                break;
+            }
+
+            let pos = nodes.len();
+            nodes.extend(next_node_layer);
+
+            node_layer = &nodes[pos..];
```

Comment:
Maybe we could merge that commit with the previous one?

Reply:
Yes I'll do that.